### PR TITLE
fortran: Fix many Fortran binding bugs

### PR DESCRIPTION
**Pull request by kawashima-fj**
@jsquyres Please review.

This PR fixes many Fortran binding issues. Please see each commit message.

### CODE REVIEW

File: ompi/mpi/fortran/mpif-h/alloc_mem_f.c [hunk 1]
```diff
@@ -63,14 +65,16 @@ OMPI_GENERATE_F77_BINDINGS (PMPI_ALLOC_MEM_CPTR,
...
 #pragma weak MPI_ALLOC_MEM_CPTR = ompi_alloc_mem_f
 #pragma weak mpi_alloc_mem_cptr = ompi_alloc_mem_f
 #pragma weak mpi_alloc_mem_cptr_ = ompi_alloc_mem_f
 #pragma weak mpi_alloc_mem_cptr__ = ompi_alloc_mem_f
 
 #pragma weak MPI_Alloc_mem_f = ompi_alloc_mem_f
 #pragma weak MPI_Alloc_mem_f08 = ompi_alloc_mem_f
+
+#pragma weak MPI_Alloc_mem_cptr_f = ompi_alloc_mem_f
+#pragma weak MPI_Alloc_mem_cptr_f08 = ompi_alloc_mem_f
```
Comment (jsquyres) [inline ompi/mpi/fortran/mpif-h/alloc_mem_f.c]: Super minor nit: in the above section, it goes:

- pragmas for alloc_mem
- pragmas for alloc_mem_f/_f08
- ...and then all the cptr pragmas

But in this section, it's a different order:
- pragmas for alloc_mem
- pragmas for cptr
- pragmas for alloc_mem_f/f08
- more pragmas for cptr

Can you make the order between these two sections be the same?

(again, I realize this is a super-picky nit -- thank you for your patience!)

**Comment by jsquyres** [comment]
@kawashima-fj Thank you for cleaning up all these issues.  Can you please PR this stuff to v2.0.0?

This PR shows how badly we need to overhaul our Fortran bindings :frowning: 

Specifically: our Fortran bindings are functionally correct, but whenever a dev adds a new MPI interface, it's super easy to leave out one of the X (where X=large) necessary Fortran implementations of that MPI interface.

A long time ago, I did the first steps of generating all the Fortran bindings at configure/make time.  Slightly less than a long time ago, I gave that code to @ggouaillardet -- @ggouaillardet have you been able to make any progress with it?  Is it something you could possibly collaborate with @kawashima-fj to make some progress on, perchance?

**Comment by kawashima-fj** [comment]
I pushed a commit to fix the order issue. Thanks.
I'll make a PR to v2.0.0 soon.

@jsquyres @ggouaillardet
I looked only `mpif-h` and `use-mpi-ignore-tkr` this time.
`use-mpi-f08` etc. may have similar issues possibly.
I also think auto-generating is desired.
Yes, possibly I can collaborate.

**Comment by jsquyres** [comment]
@kawashima-fj I think something was wrong with this PR.  Last night, I got a build failure with MTT:

```
$ ./configure "CFLAGS=-g -pipe" --enable-picky --enable-debug --enable-mpirun-prefix-by-default --disable-dlopen --disable-weak-symbols

...lots of output...

$ make -j 32

... lots of output...

  ^profile/.libs/libmpi_mpifh_pmpi.a(palloc_mem_f.o): In function `ompi_alloc_mem_cptr_f':
/home/mpiteam/scratches/community/2016-04-20cron/Wniq/mpi-install/rAXy/src/openmpi-v1.10.2-154-gfe1dc36/ompi/mpi/fortran/mpif-h/profile/palloc_mem_f.c:115:
multiple definition of `ompi_alloc_mem_cptr_f'
.libs/alloc_mem_f.o:/home/mpiteam/scratches/community/2016-04-20cron/Wniq/mpi-install/rAXy/src/openmpi-v1.10.2-154-gfe1dc36/ompi/mpi/fortran/mpif-h/alloc_mem_f.c:115:
first defined herecollect2: error: ld returned 1 exit status
make[3]: *** [libmpi_mpifh.la] Error 1
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
```

It looks like something is wrong in `alloc_mem_f.f`.

Can you investigate?

**Comment by jsquyres** [comment]
@kawashima-fj Here's the link to the MTT failure: https://mtt.open-mpi.org/index.php?do_redir=2292

**Comment by kawashima-fj** [comment]
I'll take a look.

**Comment by kawashima-fj** [comment]
@jsquyres I opened a PR to fix this issue. Please review open-mpi/ompi-release#1091.

**Pull request merged**